File: Lecture061 Queue Interview Questions/reverseFirstK.cpp
// { Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;
queue<int> modifyQueue(queue<int> q, int k);
int main() {
    int t;
    cin >> t;
    while (t-- > 0) {
        int n, k;
        cin >> n >> k;
        queue<int> q;
        while (n-- > 0) {
            int a;
            cin >> a;
            q.push(a);
        }
        queue<int> ans = modifyQueue(q, k);
        while (!ans.empty()) {
            int a = ans.front();
            ans.pop();
            cout << a << " ";
        }
        cout << endl;
    }
}// } Driver Code Ends


// User function Template for C++

// Function to reverse first k elements of a queue.
queue<int> modifyQueue(queue<int> q, int k) {
    //algo:
    //first k element stack me daaldo  and queue se hatado
    // k element stack se nikalke wapas queue me daalde 
    // fer queue k (n-k) starting k elements, wapas queue me daalde
    
    stack<int> s;
    
    for(int i = 0; i<k; i++) {
        int val = q.front();
        q.pop();
        s.push(val);
    }
    
    while(!s.empty()) {
        int val = s.top();
        s.pop();
        q.push(val);
    }
    
    int t = q.size()-k;
    
    while(t--) {
        int val = q.front();
        q.pop();
        q.push(val);
    }
    
    return q;
}

File: Lecture061 Queue Interview Questions/reverseFirstK.py
from queue import Queue

def modify_queue(q, k):
    

    stack = []
    
    for _ in range(k):
        val = q.get()
        stack.append(val)
    
    while stack:
        val = stack.pop()
        q.put(val)
    
    t = q.qsize() - k
    
    for _ in range(t):
        val = q.get()
        q.put(val)
    
    return q

# Driver code
t = int(input())
for _ in range(t):
    n, k = map(int, input().split())
    q = Queue()
    elements = list(map(int, input().split()))
    for a in elements:
        q.put(a)
    
    ans = modify_queue(q, k)
    
    while not ans.empty():
        a = ans.get()
        print(a, end=" ")
    print()






# Algorithm:
    # Push the first k elements into a stack and pop them from the queue
    # Pop elements from the stack and push them back into the queue
    # Pop the remaining (n-k) elements from the front of the queue and push them back into the queue
